### ISSUE TITLE
Bindings to #element name inside dataTemplates don't work

### ISSUE THREAD

**Issue by ahopper** [AvaloniaUI/Avalonia#1572]
Binding to element name using # from a DataTemplate does not appear to work if the element is outside of  the DataTemplate in an ItemsControl or ListBox

```
<StackPanel Width="400" Name="top">
    <TextBlock>#top.Width=</TextBlock>
    <TextBlock Text="{Binding #top.Width}"/>
    <ListBox >
      <ListBox.Items>
        dummy
      </ListBox.Items>
      <ListBox.ItemTemplate>
        <DataTemplate>
          <StackPanel Orientation="Horizontal" Gap="6">
            <TextBlock>#top.Width=</TextBlock>
            <TextBlock Text="{Binding #top.Width}"/>
            <TextBlock>$parent[3].Width=</TextBlock>
            <TextBlock Text="{Binding $parent[3].Width}"/>
          </StackPanel>

        </DataTemplate>
      </ListBox.ItemTemplate>
    </ListBox>
  </StackPanel>
```
the #top.Width fails inside the DataTemplate

**Comment by grokys** [comment]
I've investigated this a little, and the problem is occurring because each `ListBoxItem` has it's own name scope, so at the point of the binding, the `#top` control can't be found because it's in a parent name scope.

I took a look at WPF, and WPF also creates a new name scope for `ListBoxItems` so I thought this wouldn't work there either, _however_ it does! I'm not 100% sure what the mechanism here is: `ElementName` in a binding appears in this case to search parent name scopes if the name isn't found in the immediate name scope, but I'm pretty sure it doesn't do this in all cases. Need to investigate more.

**Comment by grokys** [comment]
Here's an example of this working in WPF:

```xml
<Window x:Class="WpfApp1.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:sys="clr-namespace:System;assembly=mscorlib"
        mc:Ignorable="d" Title="MainWindow" Height="450" Width="800">
    <StackPanel>
        <Button Name="foo">Button</Button>
        <ListBox>
            <ListBox.ItemsSource>
                <x:Array Type="{x:Type sys:String}">
                    <sys:String>Hello</sys:String>
                    <sys:String>World</sys:String>
                </x:Array>
            </ListBox.ItemsSource>
            <ListBox.ItemTemplate>
                <DataTemplate>
                    <StackPanel>
                        <!--<Label Name="foo" Content="{Binding}"/>-->
                        <TextBlock Text="{Binding Content, ElementName=foo}"/>
                    </StackPanel>
                </DataTemplate>
            </ListBox.ItemTemplate>
        </ListBox>
    </StackPanel>
</Window>
```

**Comment by grokys** [comment]
Ok, so this seems not just limited to `DataTemplate` - it also works for `ControlTemplates`. The rule appears to be that parent namescopes are searched _if_ they're defined in the same XAML file as far as I can see.

**Comment by kekekeks** [comment]
>UserControl defined in a separate XAML file, then the binding doesn't work!

Controls with their own XAML have isolated name scope. Same will happen if you define your DataTemplate in a separate XAML file and import via resources.

Basically ElementName only works within one XAML file, but as long it's in the same file it works freaking everywhere.

**Comment by grokys** [comment]
> Controls with their own XAML have isolated name scope.

Do you know of any documentation about this?

Controls in a name scope but have unique names, so I don't think there's a single isolated name scope with all of the controls in the xaml, because in the example above, if the `Label#foo` control is uncommented, there will be 3 controls with the name `foo`.

**Comment by kekekeks** [comment]
I *think* controls can pass through parent->child namescope boundaries if both were defined in the same XAML file.

At least this behavior makes sense.

**Comment by grokys** [comment]
Yeah that seems to be the behavior I'm seeing... Was just wondering if it was documented anywhere? I can't even find as much as a stackoverflow post that mentions it.

**Comment by grokys** [comment]
I've been looking into this some more, and I rally can't quite work out the mechanism here. It doesn't seem to actually be tied to the controls being in the same XAML file. There seem to be certain rules about when a binding will search parent namescopes.

I think the best thing to do here would be to wait until the relevant WPF code is open-sourced and take a look at what they do, then decide if we want to use the same rules or not.

**Comment by ahopper** [comment]
fixed by #2705 :+1
